crates/schema-inference/src/json_decoder.rs

```diff
@@ -0,0 +1,75 @@
+/// Largely copied from https://github.com/mxinden/asynchronous-codec/blob/master/src/codec/json.rs
+
+use std::marker::PhantomData;
+
+use bytes::{BytesMut, Buf};
+use serde::Deserialize;
+use serde_json::Value;
+use tokio_util::codec::Decoder;
+
+#[derive(Default)]
+pub struct JsonCodec<Dec = Value> {
+    dec: PhantomData<Dec>
+}
+
+impl<Dec> JsonCodec<Dec>
+where
+    for<'de> Dec: Deserialize<'de> + 'static,
+{
+    /// Creates a new `JsonCodec` with the associated types
+    pub fn new() -> JsonCodec<Dec> {
+        JsonCodec {
+            dec: PhantomData,
+        }
+    }
+}
+
+impl<Dec> Clone for JsonCodec<Dec>
+where
+    for<'de> Dec: Deserialize<'de> + 'static,
+{
+    /// Clone creates a new instance of the `JsonCodec`
+    fn clone(&self) -> JsonCodec<Dec> {
+        JsonCodec::new()
+    }
+}
+
+/// JSON Codec error enumeration
+#[derive(Debug, thiserror::Error)]
+pub enum JsonCodecError {
+    #[error("IO error")]
+    Io(#[from] std::io::Error),
+    #[error("JSON error")]
+    Json(#[from] serde_json::Error),
+}
+
+/// Decoder impl parses json objects from bytes
+impl<Dec> Decoder for JsonCodec<Dec>
+where
+    for<'de> Dec: Deserialize<'de> + 'static,
+{
+    type Item = Dec;
+    type Error = JsonCodecError;
+
+    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
```

Comment:
I actually thought that this was fine as it is, but I did a little more digging since you'd asked about it, and discovered that there is an important missing optimization. The [`Decoder` docs](https://docs.rs/tokio-util/latest/tokio_util/codec/trait.Decoder.html#tymethod.decode) seem to indicate that the `FramedRead` may not do some sub-optimal bufffer management unless the `Decoder` explicitly reserves more space in the buffer. Looking at the [framed read source](https://github.com/tokio-rs/tokio/blob/master/tokio-util/src/codec/framed_impl.rs#L216) shows that "sub-optimal" is quite an understatement! So I think we just need to grow the buffer here if we're returning `None` and we're not at EOF.

Given that we're taking on buffer management here is also making me reconsider my previous position on whether we should impose a limit on the maximum document/buffer size. It's seeming like a maximum size may be a good idea. I'll propose something rather absurdly large, like 128MiB, and we can defer making a decision on any sort of global document size limit. If for some reason a document goes over that size, then schema inference can fail :man_shrugging:

cc @jgraettinger, since he may have a more informed opinion on a max document size.

Reply:
I made it grow by 1mb every time it fails to decode a document without reaching EOF or an error